File: scala-project/src/main/scala/com/waitingforcode/SessionsGenerationJob.scala
package com.waitingforcode

import io.delta.sql.DeltaSparkSessionExtension
import org.apache.spark.sql.execution.streaming.ProcessingTimeTrigger
import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.sql.streaming.DataStreamWriter
import scopt.OParser

import java.util.TimeZone

// --kafkaBootstrapServers localhost:9094 --kafkaInputTopic visits --kafkaOutputTopic sessions --devicesTableLocation /tmp/dais2024/devices --checkpointLocation /tmp/dais24_checkpoint/
object SessionsGenerationJob {

  def main(args: Array[String]): Unit = {
    OParser.parse(SessionsGenerationJobArguments.Parser, args,
      SessionsGenerationJobArguments()) match {
      case Some(jobConfig) => {
        TimeZone.setDefault(TimeZone.getTimeZone("UTC"))
        val sparkSession = SparkSession.builder()
          .master("local[*]")
          .config("spark.sql.session.timeZone", "UTC")
          .withExtensions(new DeltaSparkSessionExtension())
          .config("spark.sql.shuffle.partitions", 2)
          .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")
          .getOrCreate()

        val inputDataStream = sparkSession.readStream
          .options(Map(
            "kafka.bootstrap.servers" -> jobConfig.kafkaBootstrapServers,
            "subscribe" -> jobConfig.kafkaInputTopic,
            "maxOffsetsPerTrigger" -> "1000"
          )).format("kafka").load()

        DevicesToDeltaConverter.convertDevicesToDeltaTable(sparkSession, jobConfig.devicesTableLocation,
          jobConfig.devicesDeltaTableLocation)
        val devicesTable = sparkSession.read.format("delta").load(jobConfig.devicesDeltaTableLocation)

        val sessionsWriter: DataStreamWriter[Row] = SessionsGenerationJobLogic.generateSessions(
          inputDataStream, devicesTable, ProcessingTimeTrigger("10 seconds")
        )

        val writerQuery = sessionsWriter.options(Map(
          "checkpointLocation" -> jobConfig.checkpointLocation,
          "kafka.bootstrap.servers" -> jobConfig.kafkaBootstrapServers,
          "topic" -> jobConfig.kafkaOutputTopic
        )).format("kafka").start()

        writerQuery.awaitTermination()
      }
    }
  }
}

case class SessionsGenerationJobArguments(kafkaBootstrapServers: String = null,
                                          kafkaInputTopic: String = null,
                                          kafkaOutputTopic: String = null,
                                          devicesTableLocation: String = null,
                                          checkpointLocation: String = null) {

  lazy val devicesDeltaTableLocation = s"${devicesTableLocation}/delta"

}
object SessionsGenerationJobArguments {
  val Parser = {
    val builder = OParser.builder[SessionsGenerationJobArguments]
    import builder._
    OParser.sequence(
      opt[String]("kafkaBootstrapServers")
        .action((kafkaBootstrapServers, c) => c.copy(kafkaBootstrapServers = kafkaBootstrapServers)),
      opt[String]("kafkaInputTopic")
        .action((kafkaInputTopic, c) => c.copy(kafkaInputTopic = kafkaInputTopic)),
      opt[String]("kafkaOutputTopic")
        .action((kafkaOutputTopic, c) => c.copy(kafkaOutputTopic = kafkaOutputTopic)),
      opt[String]("devicesTableLocation")
        .action((devicesTableLocation, c) => c.copy(devicesTableLocation = devicesTableLocation)),
      opt[String]("checkpointLocation")
        .action((checkpointLocation, c) => c.copy(checkpointLocation = checkpointLocation)),
    )
  }
}

File: python-project/visits_generator/sessions_generation_job.py
import argparse

from delta import configure_spark_with_delta_pip
from pyspark.sql import SparkSession, DataFrame

from visits_generator.devices_to_delta_converter import convert_devices_to_delta_table
from visits_generator.sessions_generation_job_logic import generate_sessions

if __name__ == '__main__':
    parser = argparse.ArgumentParser(prog='Sessionization pipeline example for DAIS 2024')
    parser.add_argument('--kafka_bootstrap_servers', required=True)
    parser.add_argument('--kafka_input_topic', required=True)
    parser.add_argument('--kafka_output_topic', required=True)
    parser.add_argument('--devices_table_location', required=True)
    parser.add_argument('--checkpoint_location', required=True)
    args = parser.parse_args()

    spark_session = configure_spark_with_delta_pip(SparkSession.builder.master("local[*]")
                                                   .config("spark.sql.shuffle.partitions", 2)
                                                   .config("spark.sql.extensions",
                                                           "io.delta.sql.DeltaSparkSessionExtension")
                                                   .config("spark.sql.catalog.spark_catalog",
                                                           "org.apache.spark.sql.delta.catalog.DeltaCatalog"),
                                                   extra_packages=['org.apache.spark:spark-sql-kafka-0-10_2.12:3.5.0']
                                                   ).getOrCreate()

    input_data_stream = (spark_session.readStream
                         .option('kafka.bootstrap.servers', args.kafka_bootstrap_servers)
                         .option('subscribe', args.kafka_input_topic)
                         .option('startingOffsets', 'LATEST')
                         .option('maxOffsetsPerTrigger', '1000')
                         .format('kafka').load())

    devices_delta_location = f'{args.devices_table_location}/delta'
    convert_devices_to_delta_table(spark_session, args.devices_table_location, devices_delta_location)
    devices_table: DataFrame = spark_session.read.format('delta').load(devices_delta_location)

    sessions_writer = generate_sessions(input_data_stream, devices_table, {'processingTime': '10 seconds'},
                                        args.checkpoint_location)

    write_query = (sessions_writer
                   .format("kafka")
                   .option('kafka.bootstrap.servers', args.kafka_bootstrap_servers)
                   .option('topic', args.kafka_output_topic)
                   .start())

    write_query.awaitTermination()
